x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+import { Observable, Subscription, from } from 'rxjs';
+import { first } from 'rxjs/operators';
+import { v4 } from 'uuid';
+import Wreck from 'wreck';
+
+import {
+  PluginInitializerContext,
+  ClusterClient,
+  Logger,
+  CoreStart,
+  CoreSetup,
+  Router,
+} from 'target/types/server';
+
+import { HttpServerSetup } from 'target/types/server/http/http_server';
+import { ProxyConfig } from './proxy_config';
+
+export enum RouteState {
+  Initializing,
+  Started,
+  Closed,
+}
+
+export interface RoutingNode {
+  type: string; // what are all the types this can be?
+  node: string;
+  state: RouteState;
+}
+
+interface LivenessNode {
+  lastUpdate: number;
+  lastChangedTime: number;
+}
+
+interface ClusterDoc {
+  nodes: {
+    [key: string]: LivenessNode;
+  };
+  routing_table: {
+    [key: string]: RoutingNode;
+  };
+}
+
+export type RoutingTable = Map<string, RoutingNode>;
+type NodeList = Map<string, LivenessNode>;
+
+// convert a map<string, V> to a plain javascript object of {[key:string]: val}
+function map2obj<V>(map: Map<string, V>): { [key: string]: V } {
+  return [...map].reduce((a, [k, v]) => Object.assign(a, { [k]: v }), {});
+}
+
+export class ProxyService {
+  private updateInterval: number = 1000 * 15;
+  private timeoutThreshold: number = 1000 * 30;
+  private port: number = 5602;
+  private timer: null | NodeJS.Timeout = null;
+  private configSubscription?: Subscription;
+  private routingTable: RoutingTable = new Map();
+  private elasticsearch?: Observable<ClusterClient>;
+  private proxyServer?: HttpServerSetup;
+  private nodeName: string;
+
+  private readonly proxyIndex = '.kibana';
+  private readonly proxyDoc = 'proxy-resource-list';
+  private readonly httpsAgent: HTTPSAgent;
+  private readonly httpAgent: HTTPAgent;
+  private readonly allowUnauthAgent: HTTPAgent;
+  private readonly wreck: typeof Wreck;
+  private readonly log: Logger;
+  private readonly config$: Observable<ProxyConfig>;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.nodeName = v4();
+    this.config$ = initializerContext.config.create<ProxyConfig>();
+    this.log = initializerContext.logger.get('proxy');
+    this.httpsAgent = new HTTPSAgent({ keepAlive: true });
+    this.httpAgent = new HTTPAgent({ keepAlive: true });
+    this.allowUnauthAgent = new HTTPAgent({ keepAlive: true });
+    this.wreck = Wreck.defaults({
+      agent: {
+        https: this.httpsAgent,
+        http: this.httpAgent,
+        httpsAllowUnauthorized: this.allowUnauthAgent,
+      },
+    });
+  }
+
+  public async setup(core: CoreSetup) {
+    // initialize cluster information
+    // we need to get this kibana instance's ip address
+    // it's hostname and determine which port to listen on
+    // should liveness timing be part of configuration?o
+    // mikhael -- talk about changing http service to allow multiple services
+    this.configSubscription = this.config$.subscribe(config => {
+      this.setConfig(config);
```

Comment:
not sure it's necessary
```js
const config = await this.config$.pipe(first()).toPromise();
```
should always provide the latest config values

Reply:
This is probably such a common pattern that we may want to provide a function that returns this from Core

Reply:
i copied this from core -- it seems like it allows the config to update without having to restart the server?  there aren't many options though i think you should be able to change without a restart?

Reply:
Although we haven't implemented config changes without restart in Kibana, the `config$` is an observable so that it would be possible to one day support it. So to subscribe to `config$` is the most future proof way to consume configuration :+1:

However, the rest of the proxy implementation doesn't react to changes in the configuration (setConfig doesn't update the underlying http server's port for instance). I also don't think it's a requirement for the first version of the proxy plugin, so perhaps it's clearer if the subscription is removed and we just use 
```
const config = await this.config$.pipe(first()).toPromise();
this.setConfig(config);
```

Until the plugin actually supports reacting to config changes.